performance-tests/bench/tcp_latency/tcp_latency.cpp

```diff
@@ -0,0 +1,251 @@
+#include "BenchC.h"
+#include "PropertyStatBlock.h"
+
+#include <json_conversion.h>
+
+#include "ace/OS_NS_string.h"
+#include "ace/OS_NS_sys_wait.h"
+#include "ace/Thread.h"
+#include "ace/Thread_Manager.h"
+#include "ace/SOCK_Stream.h"
+#include "ace/SOCK_Connector.h"
+#include "ace/SOCK_Acceptor.h"
+#include "ace/Log_Msg.h"
+#include "ace/Time_Value.h"
+#include "ace/OS_NS_unistd.h"
+#include "ace/OS_NS_time.h"
+#include "ace/Get_Opt.h"
+
+#include <numeric>
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <cmath>
+#include <vector>
+
+const int max_buffer_size = 32000;
+
+ACE_INET_Addr server_addr;
+int frequency = 100; // messages per second
+int message_size = 0;
+int test_duration = 120; // seconds
+int server_port = 0;
+std::string report_file_name;
+
+struct DataSample {
+  ACE_hrtime_t sending_time_;
+  char padding_[max_buffer_size - sizeof(ACE_hrtime_t) ];
+};
+
+static void *
+sender (void *arg)
+{
+  ACE_SOCK_Stream& cli_stream = *reinterpret_cast<ACE_SOCK_Stream*>(arg);
+
+  ACE_Time_Value sleep_time (0, 1000000/frequency);
+  int counter = frequency * test_duration;
+  DataSample data;
+
+  while (counter--) {
+    data.sending_time_ = ACE_OS::gethrtime();
+    if (cli_stream.send_n (&data, message_size) == -1)
+    {
+
+      ACE_ERROR((LM_ERROR,
+                 ACE_TEXT("(%P|%t) %p\n"), ACE_TEXT("cli_stream.send_n")));
+      return 0;
+    }
+    ACE_OS::sleep(sleep_time);
+  }
+
+  cli_stream.close_writer();
+
+  return 0;
+}
+
+static void
+receiver(ACE_SOCK_Stream& cli_stream)
+{
+  ACE_INET_Addr peer_addr;
+  DataSample data;
+  ssize_t rcv_cnt = 1;
+
+  Bench::WorkerReport report{};
+  report.process_report.participants.length(1);
+  report.process_report.participants[0].subscribers.length(1);
+  report.process_report.participants[0].subscribers[0].datareaders.length(1);
+  Builder::PropertySeq& properties = report.process_report.participants[0].subscribers[0].datareaders[0].properties;
+
+  const size_t buffer_size = static_cast<size_t>(1.1 * frequency * test_duration);
+
+  auto round_trip_latency_stat_block = std::make_shared<Bench::PropertyStatBlock>(properties, "round_trip_latency", buffer_size);
```

Comment:
Are we allowed to use auto in OpenDDS, or is this just because bench is run specifically on a more modern machine with auto?

Reply:
Bench requires C++11

Reply:
Bench uses C++11 so `auto` is available.